processes: add `cgroup_path` column on Linux

artemist-work opened the pull request:
This will read the full cgroup path out of /proc for each process in the process table. I've tested this with cgroup v2 (unified hierarchies) but have not yet tested it with cgroup v1 (where each group type has a separate hierarchy). On cgroup v1 it should report the first cgroup it's in but that may not be consistently the same type. Proper cgroup v1 support would require a different schema as each process can be in 13 different hierarchies.

This does involve several string copies but I couldn't find a clean way of doing it with fewer.

directionless commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +58,32 @@ inline std::string readProcCMDLine(const std::string& pid) {
+  auto attr = getProcAttr("cgroup", pid);
+
+  std::string content;
+  readFile(attr, content);
```
Naive question... Is there a performance impact from there? I think we use the process table for a variety of internal monitoring, right? And this would add a proc read per process per query.

artemist-work replied:
Yes, this will take more time in order to do a proc read. The processes table already has several other proc reads. I'd want some way to lazily execute the code but without that we're stuck between "an extra proc read for every process" or "another table with a join". Looking at [a blog post about syscall time](https://gms.tf/on-the-costs-of-syscalls.html) it seems like we should be adding on the order of 2µs per process, though I haven't tested this.

alessandrogario replied:
Thanks to the latest changes applied, I think the potential impact on performances has been reduced. Here's a (rough) quick test with `.timer on`:

processes
 * 21463 processes on the host
 * 2046 of which are from a container
 * runtime used: libpod
 * 10 runs on each branch

cgroup paths had the following values:
 * On Fedora, most processes will return cgroup_path as '/'
 * 66 had `/system.slice/%`
 * 93 had `/user.slice/%`
 * 2046 had `/machine.slice/libpod-%`

offsets
 * real +9.48%
 * user +8.82%
 * sys +12.40%

Perf should be able to provide better readings (and a diff) if required

artemist-work replied:
Latest patch should have no measurable perf impact if `cgroup_path` is not used

directionless replied:
Thanks for updating this

alessandrogario commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +58,32 @@ inline std::string readProcCMDLine(const std::string& pid) {
+    return "";
+  }
+
+  return boost::join(std::vector(fields.begin() + 2, fields.end()), ":");
```
Is this `boost::split` + `boost::join` returning a substr from the second `:` character until the end?

artemist-work replied:
Yes. I spent a while looking for a split function where I could tell it to split into exactly 3 but I couldn't find one

alessandrogario reviewed (changes requested):
This is a great feature, and will make the `processes` table incredibly useful for hosts with running containers!

artemist-work commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +58,32 @@ inline std::string readProcCMDLine(const std::string& pid) {
+  if (auto pos = content.find_first_of('\n'); pos != std::string::npos) {
+    content.erase(pos);
+  }
+  boost::algorithm::trim_right_if(content,
```
This should erase everything after (and including) the first new line. It seems to work in [compiler explorer](https://godbolt.org/z/evMP7fP3v) and I haven't been getting spurious newlines when testing

artemist-work commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +58,32 @@ inline std::string readProcCMDLine(const std::string& pid) {
   return content;
 }
 
+inline std::string readProcCgroup(const std::string& pid) {
```
I could definitely extract it, I guess I would call the parse function with the output of readFile.
I'll try that in another commit and add some tests

artemist-work commented:
Okay, I added some unit tests tests and cleaned up `/proc/<pid>/cgroup` parsing to use fewer copies. It's still a bit strange since C++ functions always seem to do precisely what you don't want them to do.

alessandrogario commented on osquery/tables/system/linux/processes.cpp:
```diff
@@ -56,6 +59,35 @@ inline std::string readProcCMDLine(const std::string& pid) {
+  auto attr = getProcAttr("cgroup", pid);
+
+  std::string content;
+  readFile(attr, content);
```
```suggestion
  if (!readFile(attr, content).ok()) {
    return {};
  }

```

artemist-work commented:
My latest patch should remove the time penalty if cgroup is not used but build seems to be failing. I can't pinpoint the exact bug and I seem to not have permission to rerun
